actors/miner/src/beneficiary.rs

```diff
@@ -0,0 +1,87 @@
+use fvm_ipld_encoding::tuple::*;
+use fvm_ipld_encoding::Cbor;
+use fvm_shared::address::Address;
+use fvm_shared::bigint::bigint_ser;
+use fvm_shared::clock::ChainEpoch;
+use fvm_shared::econ::TokenAmount;
+use std::cmp::max;
+
+#[derive(Debug, PartialEq, Serialize_tuple, Deserialize_tuple)]
+pub struct BeneficiaryTerm {
+    // Quota: The total amount the current beneficiary can withdraw. Monotonic, but reset when beneficiary changes.
+    #[serde(with = "bigint_ser")]
+    pub quota: TokenAmount,
+    // UsedQuota: The amount of quota the current beneficiary has already withdrawn
+    #[serde(with = "bigint_ser")]
+    pub used_quota: TokenAmount,
+    // Expiration: The epoch at which the beneficiary's rights expire and revert to the owner
+    pub expiration: ChainEpoch,
+}
+
+impl Cbor for BeneficiaryTerm {}
+
+impl BeneficiaryTerm {
+    pub fn default() -> BeneficiaryTerm {
+        BeneficiaryTerm {
+            quota: TokenAmount::default(),
+            expiration: 0,
+            used_quota: TokenAmount::default(),
+        }
+    }
+
+    pub fn new(
+        quota: TokenAmount,
+        used_quota: TokenAmount,
+        expiration: ChainEpoch,
+    ) -> BeneficiaryTerm {
+        BeneficiaryTerm { quota, expiration, used_quota }
+    }
+
+    // IsUsedUp check whether beneficiary has use up all quota
+    pub fn is_used_up(&self) -> bool {
+        self.used_quota >= self.quota
+    }
+
+    // IsExpire check if the beneficiary is within the validity period
+    pub fn is_expire(&self, cur: ChainEpoch) -> bool {
+        self.expiration <= cur
+    }
+
+    // Available get the amount that the beneficiary has not yet withdrawn
+    pub fn available(&self, cur: ChainEpoch) -> TokenAmount {
+        // Return 0 when the usedQuota > Quota for safe
+        if self.is_expire(cur) {
+            TokenAmount::default()
+        } else {
+            max(self.quota.clone() - self.used_quota.clone(), TokenAmount::default())
```

Comment:
I don't have heaps of experience with this myself, but would guess there's some way to do this without  so many clones (I expect the method to clone internally). Try `self.quota.sub(self.used_quota)`?

Reply:
Yes, you should be able to take references:

```rust
(&self.quota).sub(&self.used_quota).max(TokenAmount::default())
```

Or:

```rust
use std::borrow::Borrow;
self.quota.borrow().sub(&self.used_quota).max(TokenAmount::default())
```

(There may be neater ways to clamp at 0; I'm pretty sure they are).

---

Furthermore, you can get rid of the if adopting a more functional style, but this may harm readability:

```rust
self.is_expire(cur)
    .then(|| (&self.quota).sub(&self.used_quota).max(TokenAmount::default()))
    .unwrap_or_default()
```

---

Wouldn't you want to use a `checked_sub` here to prevent overflows, though?

Reply:
Furthermore, you can get rid of the if adopting a more functional style, but this may harm readability:

```rust
self.is_expire(cur)
    .then(|| (&self.quota).sub(&self.used_quota).max(TokenAmount::default()))
    .unwrap_or_default()
```

Reply:
Wouldn't you want to use a `checked_sub` here to prevent overflows, though?

Reply:
@raulk I had commented earlier that the BigInt under TokenAmount never overflows, so @hunjixin removed all the checked arithmetic for readability.

Reply:
unwrap_or_default or unwrap?, this code should never fail, i think better to use unwrap here

Reply:
hmmm，i remove is_usedup and is_expire, just keep available, user could use this function to check this beneficiary is efficent or not. code will be cleaner

Reply:
Yes that looks good to me (and I prefer the readability of the explicit `if` block).